shepherd/shepherd/shepherd.py

```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
-                    error = ErrorModel({
-                        'message': 'Failed to start sheep for this job ({})'.format(str(sce))
-                    })
-
-                    logging.error('Sheep `%s` encountered error when processing job `%s`: %s', sheep_id, job_id, error.message)
-                    await self._report_job_failed(job_id, error, sheep)
-                    continue
-                except Exception as ex:
-                    error = ErrorModel({
-                        'message': '`{}` thrown when starting sheep `{}` for job `{}`'.format(str(ex), sheep_id, job_id),
-                        'exception_type': str(type(ex)),
-                        'exception_traceback': str(traceback.format_tb(ex.__traceback__))
-                    })
-
-                    await self._report_job_failed(job_id, error, sheep)
-                    logging.exception("Error encountered when starting sheep `%s` for job `%s`", sheep_id, job_id)
+                    await self._prepare_job_dir_futures[job_id]
+                    del self._prepare_job_dir_futures[job_id]
+                except KeyError:
+                    logging.critical(f"Task that should have pulled data for job `{job_id}` mysteriously disappeared")
                     continue
 
-            # send the InputMessage to the sheep
-            sheep.in_progress.add(job_id)
-            logging.info('Sending InputMessage for job `%s` on `%s`', job_id, sheep_id)
-            await Messenger.send(sheep.socket, InputMessage(dict(job_id=job_id, io_data_root=sheep.sheep_data_root)))
-
-            # notify the queue that the task is done
-            sheep.jobs_queue.task_done()
+                # update the job status
+                status = self._job_status[job_id]
+                status.status = JobStatus.PROCESSING
+                status.processing_started_at = datetime.utcnow()
+                await self._job_status_update_queue.enqueue_task(self._storage.set_job_status(job_id, status.copy()))
+
+                # (re)start the sheep if needed
+                model = status.model
+                if model.name != sheep.model_name or model.version != sheep.model_version or not sheep.running:
+                    logging.info('Job `%s` requires model `%s:%s` on `%s`', job_id, model.name, model.version, sheep_id)
+                    reconfiguration_successful = await self._reconfigure_sheep(sheep, job_id, model)
+
+                    if not reconfiguration_successful:
+                        continue
+
+                # send the InputMessage to the sheep
+                sheep.in_progress.add(job_id)
+                logging.info('Sending InputMessage for job `%s` on `%s`', job_id, sheep_id)
+                await Messenger.send(
+                    sheep.socket, InputMessage(dict(job_id=job_id, io_data_root=sheep.sheep_data_root))
+                )
+
+                # notify the queue that the task is done
```

Comment:
This is weird wording (line 303 as well), the job is certainly not done, it was just sent for processing